std/string.d

```diff
@@ -2937,26 +2937,96 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
 {
     static import std.ascii;
     static import std.uni;
-    import std.utf : decodeFront;
 
-    while (!input.empty)
+    static if (is(Unqual!(ElementEncodingType!Range) == dchar)
+        || is(Unqual!(ElementEncodingType!Range) == wchar))
     {
-        auto c = input.front;
-        if (std.ascii.isASCII(c))
+        // Decoding is never needed for dchar. It happens not to be needed
+        // here for wchar because no whitepace is outside the basic
+        // multilingual plane meaning every whitespace character is encoded
+        // with a single wchar and due to the design of UTF-16 those wchars
+        // will not occur as part of the encoding of multi-wchar codepoints.
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
         {
-            if (!std.ascii.isWhite(c))
-                break;
-            input.popFront();
+            size_t i = 0;
+            for (const size_t end = input.length; i < end; ++i)
+            {
+                if (!std.uni.isWhite(input[i]))
+                    break;
+            }
+            input = input[i .. $];
+            return input;
         }
         else
         {
-            auto save = input.save;
-            auto dc = decodeFront(input);
-            if (!std.uni.isWhite(dc))
-                return save;
+            while (!input.empty)
+            {
+                if (!std.uni.isWhite(input.front))
+                    break;
+                input.popFront();
+            }
+            return input;
+        }
+    }
+    else
+    {
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
+        {{
+            // ASCII optimization for dynamic arrays & similar.
+            size_t i = 0;
+            for (const size_t end = input.length; i < end; ++i)
```

Comment:
The suggested modification in the other branch makes this cleaner as well.

Reply:
Hm... I was thinking this too could be rewritten:

```D
foreach (i; 0 .. input.length)
{
    auto c = input[i];
    if (c >= 0x80)
    {
        input = input[i .. $];
        goto NonAsciiPath;
    }
    if (!std.ascii.isWhite)
        return input[i .. $];
}
return input[$ .. $];
NonAsciiPath:
```
While it doesn't get rid of the goto, the loop looks cleaner to me.

Reply:
The rewrite makes the ASCII loop slower while avoiding a call to `_d_arraybounds`. When processing `"\n\t\v\rhello world\n\t\v\r"`, with `ldc2 -O2` the function overall takes ~5% longer and with `dmd -O -inline` it takes ~30% longer. Obviously when `-release` is added and no bounds checks are performed the comparison gets worse.

Reply:
This has to be a cache line thing, because the major work done in the loop (calling isWhite and checking if c is non-ascii) is identical in both cases. In any case, it's only marginally more readable, so the original is fine.